designs/2019-move-to-async-api/README.md

```diff
@@ -0,0 +1,420 @@
+    resolvePluginsRelativeTo = cwd,
+    rulePaths = [],
+    rules = null,
+    useEslintrc = true,
+    ...unknownOptions
+  } = {}) {
+    // Throws on unknown options
+    {
+      const keys = Object.keys(unknownOptions)
+      if (keys.length >= 1) {
+        //...
+      }
+    }
+    // Throws on the invalid value of options
+    if (typeof allowInlineConfig !== "boolean") {
+      // ...
+    }
+    // ...
+
+    // Initialize CLIEngine because this is a tiny wrapper.
+    const engine = (this._cliEngine = new CLIEngine({
+      allowInlineConfig,
+      baseConfig,
+      cache,
+      cacheLocation,
+      configFile,
+      cwd,
+      envs,
+      extensions,
+      fix,
+      fixTypes,
+      globInputPaths,
+      globals,
+      ignore,
+      ignorePath,
+      ignorePattern,
+      parser,
+      parserOptions,
+      plugins,
+      reportUnusedDisableDirectives,
+      resolvePluginsRelativeTo,
+      rulePaths,
+      rules,
+      useEslintrc,
+    }))
+
+    // Apply `pluginImplementations` option.
+    if (pluginImplementations) {
+      for (const [id, definition] of Object.entries(pluginImplementations)) {
+        engine.addPlugin(id, definition)
+      }
+    }
+  }
+}
+```
+
+</details>
+
+#### § The `executeOnFiles()` method
+
+This method returns an object that has two methods `then()` and `[Symbol.asyncIterator]()`, so we can use the returned object with `await` expression and `for-await-of` statement.
+
+- If you used the returned object with `for-await-of` statement, it iterates the lint result of each file in random order. This way yields each lint result immediately. Therefore you can print the results in streaming, or print progress state. ESLint may spend time to lint files (for example, ESLint needs about 20 seconds to lint our codebase), to print progress state will be useful.
+
+  ```js
+  const { ESLint } = require("eslint")
+  const eslint = new ESLint()
+
+  for await (const result of eslint.executeOnFiles(patterns)) {
+    print(result)
+  }
+  ```
+
+- If you used the returned object with `await` expression, it returns the array that contains all lint results in random order. This way waits for all lint results then returns it.
+
+  ```js
+  const { ESLint } = require("eslint")
+  const eslint = new ESLint()
+
+  const results = await eslint.executeOnFiles(patterns)
+  // Optionally you can sort the results.
+  // results.sort(ESLint.compareResultsByFilePath)
+  print(results)
+  ```
+
+Either way, we can support "linting in parallel", loading configs/plugins with `import()`, and other stuff that needs asynchronous logic.
+
+##### Move the `usedDeprecatedRules` property
+
+The returned object of `CLIEngine#executeOnFiles()` has the `usedDeprecatedRules` property that includes the deprecated rule IDs which the linting used. But the location doesn't fit the use with `for-await-of` statement. Therefore, this RFC moves the `usedDeprecatedRules` property to each lint result.
```

Comment:
>  I guess it will possibly differ if different files are exposed to different rule sets?

Exactly. We have cascading config files and glob-based configurations, so used deprecated rules may be different per file.

This change is just needed because we may get the used deprecated rules asynchronously.